src/fs/protocol/private/index.ts

```diff
@@ -0,0 +1,79 @@
+import { Maybe } from "../../../common"
+import ipfs, { CID } from "../../../ipfs"
+import MMPT from "./mmpt"
+import { DecryptedNode, PrivateAddResult } from './types'
+import * as check from './types/check'
+import * as namefilter from './namefilter'
+import { BareNameFilter, PrivateName } from './namefilter'
+import * as basic from '../basic'
+
+export const addNode = async (mmpt: MMPT, node: DecryptedNode, key: string): Promise<PrivateAddResult> => {
+  const { cid, size } = await basic.putEncryptedFile(node, key)
+  const filter = await namefilter.addRevision(node.bareNameFilter, key, node.revision)
+  const name = await namefilter.toPrivateName(filter)
+  await mmpt.add(name, cid)
+  return { cid, name, key, size }
+}
+
+export const readNode = async (cid: CID, key: string): Promise<DecryptedNode> => {
+  const content = await ipfs.encoded.catAndDecode(cid, key)
+  if(!check.isDecryptedNode(content)){
+    throw new Error(`Could not parse a valid filesystem object, ${cid}`)
+  }
+  return content
+}
+
+export const getByName = async (mmpt: MMPT, name: PrivateName, key: string): Promise<Maybe<DecryptedNode>> => {
+  const cid = await mmpt.get(name)
+  if(cid === null) return null
+  return getByCID(mmpt, cid, key)
+}
+
+export const getByCID = async (mmpt: MMPT, cid: CID, key: string): Promise<DecryptedNode> => {
+  const node = await readNode(cid, key)
+  const latest = await findLatestRevision(mmpt, node.bareNameFilter, key, node.revision)
+  return latest?.cid
+    ? readNode(latest?.cid, key)
+    : node
+}
+
+type Revision = {
+  cid: CID
+  name: PrivateName
+}
+
+export const findLatestRevision = async (mmpt: MMPT, bareName: BareNameFilter, key: string, lastKnownRevision: number): Promise<Maybe<Revision>> => {
+  let lowerBound = lastKnownRevision, upperBound = null
+  let i = 1
+  let lastRevision: Maybe<Revision> = null
+  while(upperBound === null){
+    const toCheck = lastKnownRevision + Math.pow(i, 2)
```

Comment:
I think this exponent is actually backwards. We want powers of two, not squares. The order is confusing, but from MDN:

```js
console.log(Math.pow(4, 0.5));
// expected output: 2
```

Reply:
ah yup that's totally right :+1: